src/codegen/cpu.rs

```diff
@@ -129,6 +87,101 @@ impl<'a> X86CodeGenerator<'a> {
         out.write_str(&main_body_str)
     }
 
+    pub fn emit_main<W: Write>(
+        &mut self,
+        top_arg_tensors: &'a [Rc<Tensor<Tgt>>],
+        out: &mut W,
+    ) -> fmt::Result {
+        let mut main_body_str = String::new();
+        let mut depth = 0_usize;
+
+        writeln!(main_body_str, "int main() {{")?;
+        depth += 1;
+        for (i, tensor) in top_arg_tensors.iter().enumerate() {
+            let c_buffer = self.name_env.get(tensor).unwrap();
+            writeln!(
+                main_body_str,
+                "{}{} buf{}[{}] __attribute__((aligned (128)));",
+                indent(depth),
+                c_type(c_buffer.dtype()),
+                i, // we do not use c_buffer.name().unwrap()
+                c_buffer.size().unwrap(),
+            )?;
+        }
+        writeln!(main_body_str)?;
+
+        // Initialize the input tensors
+        writeln!(
+            main_body_str,
+            "{}// Initialize the input tensors.",
+            indent(depth)
+        )?;
+        for (i, tensor) in top_arg_tensors
+            .iter()
+            .enumerate()
+            .take(top_arg_tensors.len() - 1)
+        {
+            let c_buffer = self.name_env.get(tensor).unwrap();
+            writeln!(
+                main_body_str,
+                "{}for (int i = 0; i < {}; i++) {{",
+                indent(depth),
+                c_buffer.size().unwrap(),
+            )?;
+            depth += 1;
+            writeln!(
+                main_body_str,
+                "{}buf{}[i] = i;",
+                indent(depth),
+                i, // we do not use c_buffer.name().unwrap()
+            )?;
+            depth -= 1;
+            writeln!(main_body_str, "{}}}", indent(depth))?;
+        }
+        writeln!(main_body_str)?;
+
+        // Emit the kernel call
+        write!(main_body_str, "{}kernel(", indent(depth))?;
+        for i in 0..top_arg_tensors.len() {
+            write!(
+                main_body_str,
+                "&buf{}[0]{}",
+                i, // we do not use c_buffer.name().unwrap()
+                if i < top_arg_tensors.len() - 1 {
+                    ", "
+                } else {
+                    ");\n"
+                },
+            )?;
+        }
+        writeln!(main_body_str)?;
```

Comment:
This can be done more simply—and with automatic switching between stack, heap, etc. according to size—by using the CBuffer abstraction as follows:

```suggestion
        // Allocate a buffer for each Impl parameter.
        let mut buffers = Vec::with_capacity(top_arg_tensors.len());
        for kernel_argument in top_arg_tensors {
            let spec = kernel_argument.spec();
            let buf = self.make_buffer(
                spec.dim_sizes(),
                spec.vector_size(),
                spec.dtype(),
                spec.level(),
            );
            buf.emit(&mut main_body_str, true, depth)?;
            buffers.push(buf);
        }

        // Emit the kernel call, passing pointers to the Impl function.
        write!(main_body_str, "\n{}kernel(", indent(depth))?;
        for (i, a) in buffers.iter().enumerate() {
            write!(
                main_body_str,
                "{}{}",
                self.c_index_ptr(a, &0i32.into(), None),
                if i < top_arg_tensors.len() - 1 {
                    ", "
                } else {
                    ");\n"
                },
            )?;
        }
        writeln!(main_body_str)?;
```

This does allocate the odd-looking, unique names we use everywhere else, but that's a separate issue we should solve more generally.

Does this work?

Additionally, we should probably call `emit_free` on these at the end of `main`.

Reply:
It appears this emits incorrect C code:

<details>
  <summary>Original</summary>

```console

---------------------------------------------------------------------------------------------------------
/* Matmul((4×4, u32), (4×4, u32), (4×4, u32), serial) */
alloc aa: (4×4, u32, L1) <- #1
  /* Matmul((4×4, u32), (4×4, u32, L1), (4×4, u32), serial) */
  tile (ab: (2×4, u32) <-[0, 2]- #0, ac: (2×4, u32) <-[0, 1]- #2)
    /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32), serial) */
    alloc ad: (2×4, u32, L1) <- ac
      /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32, L1), serial) */
      alloc ae: (2×4, u32, L1) <- ab
        /* Matmul((2×4, u32, L1), (4×4, u32, L1), (2×4, u32, L1), serial) */
        alloc af: (2×4, u32, RF)
          /* Move((2×4, u32, L1), (2×4, u32, RF), serial) */
          tile (ag: (1×1, u32, L1, ua) <-[0, 1]- ae, ah: (1×1, u32, RF, ua) <-[0, 1]- af)
            /* Move((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial) */
            ValueAssign(ag, ah)
          /* Matmul((2×4, u32, RF), (4×4, u32, L1), (2×4, u32, L1), serial) */
          tile (ai: (1×4, u32, RF, ua) <-[0, 2]- af, aj: (1×4, u32, L1, ua) <-[0, 1]- ad)
            /* Matmul((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, L1, ua), serial) */
            alloc ak: (1×4, u32, VRF, 4)
                /* Zero((1×4, u32, VRF, 4), serial) */
                VectorZero(ak)
                /* MatmulAccum((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, VRF, 4), serial) */
                tile (al: (1×1, u32, RF, ua) <-[0, 1]- ai, am: (1×4, u32, L1, ua) <-[1, 2]- aa)
                  /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, L1, ua), (1×4, u32, VRF, 4), serial) */
                  alloc an: (1×4, u32, VRF, 4)
                    /* Move((1×4, u32, L1, ua), (1×4, u32, VRF, ua, 4), serial) */
                    VectorAssign(am, an)
                    /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, VRF, 4), (1×4, u32, VRF, 4), serial) */
                    BroadcastVecMult(al, an, ak)
              /* Move((1×4, u32, VRF, 4), (1×4, u32, L1), serial) */
              VectorAssign(ak, aj)

#include <inttypes.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <arm_neon.h>
typedef uint32_t vui4 __attribute__ ((vector_size (4 * sizeof(uint32_t))));

__attribute__((noinline))
void kernel(
  uint32_t *restrict aa, 
  uint32_t *restrict ab, 
  uint32_t *restrict ac
) {
  for (int ad = 0; ad < 2; ad++) {
    uint32_t ae[8] __attribute__((aligned (128)));
    for (int af = 0; af < 2; af++) {
    for (int ag = 0; ag < 4; ag++) {
      ae[4 * af + ag] = aa[8 * ad + 4 * af + ag];
    }
    }
    for (int ah = 0; ah < 2; ah++) {
      vui4 aj;
      aj *= 0;  /* VectorZero */
      for (int ak = 0; ak < 4; ak++) {
        vui4 am;
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
        *(uint32x4_t *)(&aj) += ae[4 * ah + ak] * (*(uint32x4_t *)(&am)); /* BroadcastVecMult */
      }
      *(uint32x4_t *)(ac + 8 * ad + 4 * ah) = (*(uint32x4_t *)(&aj));  /* VectorAssign */
    }
  }
}

int main() {
  uint32_t buf0[16] __attribute__((aligned (128)));
  uint32_t buf1[16] __attribute__((aligned (128)));
  uint32_t buf2[16] __attribute__((aligned (128)));

  // Initialize the input tensors.
  for (int i = 0; i < 16; i++) {
    buf0[i] = i;
  }
  for (int i = 0; i < 16; i++) {
    buf1[i] = i;
  }

  kernel(&buf0[0], &buf1[0], &buf2[0]);

  // Print the output.
  for (int i = 0; i < 16; i++) {
    printf("%d ", buf2[i]);
  }

  return 0;
}

Output: 56 62 68 74 152 174 196 218 248 286 324 362 344 398 452 506 

```
</details>

<details>
  <summary>Proposed Code</summary>

```console

---------------------------------------------------------------------------------------------------------
/* Matmul((4×4, u32), (4×4, u32), (4×4, u32), serial) */
alloc aa: (4×4, u32, L1) <- #1
  /* Matmul((4×4, u32), (4×4, u32, L1), (4×4, u32), serial) */
  tile (ab: (2×4, u32) <-[0, 2]- #0, ac: (2×4, u32) <-[0, 1]- #2)
    /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32), serial) */
    alloc ad: (2×4, u32, L1) <- ac
      /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32, L1), serial) */
      alloc ae: (2×4, u32, L1) <- ab
        /* Matmul((2×4, u32, L1), (4×4, u32, L1), (2×4, u32, L1), serial) */
        alloc af: (2×4, u32, RF)
          /* Move((2×4, u32, L1), (2×4, u32, RF), serial) */
          tile (ag: (1×1, u32, L1, ua) <-[0, 1]- ae, ah: (1×1, u32, RF, ua) <-[0, 1]- af)
            /* Move((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial) */
            ValueAssign(ag, ah)
          /* Matmul((2×4, u32, RF), (4×4, u32, L1), (2×4, u32, L1), serial) */
          tile (ai: (1×4, u32, RF, ua) <-[0, 2]- af, aj: (1×4, u32, L1, ua) <-[0, 1]- ad)
            /* Matmul((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, L1, ua), serial) */
            alloc ak: (1×4, u32, VRF, 4)
                /* Zero((1×4, u32, VRF, 4), serial) */
                VectorZero(ak)
                /* MatmulAccum((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, VRF, 4), serial) */
                tile (al: (1×1, u32, RF, ua) <-[0, 1]- ai, am: (1×4, u32, L1, ua) <-[1, 2]- aa)
                  /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, L1, ua), (1×4, u32, VRF, 4), serial) */
                  alloc an: (1×4, u32, VRF, 4)
                    /* Move((1×4, u32, L1, ua), (1×4, u32, VRF, ua, 4), serial) */
                    VectorAssign(am, an)
                    /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, VRF, 4), (1×4, u32, VRF, 4), serial) */
                    BroadcastVecMult(al, an, ak)
              /* Move((1×4, u32, VRF, 4), (1×4, u32, L1), serial) */
              VectorAssign(ak, aj)

#include <inttypes.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <arm_neon.h>
typedef uint32_t vui4 __attribute__ ((vector_size (4 * sizeof(uint32_t))));

__attribute__((noinline))
void kernel(
  uint32_t *restrict aa, 
  uint32_t *restrict ab, 
  uint32_t *restrict ac
) {
  for (int ad = 0; ad < 2; ad++) {
    uint32_t ae[8] __attribute__((aligned (128)));
    for (int af = 0; af < 2; af++) {
    for (int ag = 0; ag < 4; ag++) {
      ae[4 * af + ag] = aa[8 * ad + 4 * af + ag];
    }
    }
    for (int ah = 0; ah < 2; ah++) {
      vui4 aj;
      aj *= 0;  /* VectorZero */
      for (int ak = 0; ak < 4; ak++) {
        vui4 am;
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
        *(uint32x4_t *)(&aj) += ae[4 * ah + ak] * (*(uint32x4_t *)(&am)); /* BroadcastVecMult */
      }
      *(uint32x4_t *)(ac + 8 * ad + 4 * ah) = (*(uint32x4_t *)(&aj));  /* VectorAssign */
    }
  }
}

int main() {
  uint32_t *restrict an;
  posix_memalign((void **)&an, 128, 16*sizeof(uint32_t));
  memset(an, 0, 16*sizeof(uint32_t));
  uint32_t *restrict ao;
  posix_memalign((void **)&ao, 128, 16*sizeof(uint32_t));
  memset(ao, 0, 16*sizeof(uint32_t));
  uint32_t *restrict ap;
  posix_memalign((void **)&ap, 128, 16*sizeof(uint32_t));
  memset(ap, 0, 16*sizeof(uint32_t));

  kernel(an + 0, ao + 0, ap + 0);

  // Print the output.
  for (int i = 0; i < 16; i++) {
    printf("%d ", buf2[i]);
  }

  return 0;
}

Error: Clang exited with exit status: 1
/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp7yEmGa/main.c:28:50: warning: incompatible pointer types assigning to 'const unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:8756:47: note: expanded from macro 'vld1q_u32'
  __ret = (uint32x4_t) __builtin_neon_vld1q_v(__p0, 50); \
                                              ^~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26817:21: note: expanded from macro 'vst1q_u32'
  uint32x4_t __s1 = __p1; \
                    ^~~~
/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp7yEmGa/main.c:28:19: warning: incompatible pointer types assigning to 'unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                  ^~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26818:26: note: expanded from macro 'vst1q_u32'
  __builtin_neon_vst1q_v(__p0, (int8x16_t)__s1, 50); \
                         ^~~~
/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp7yEmGa/main.c:51:19: error: use of undeclared identifier 'buf2'
    printf("%d ", buf2[i]);
                  ^
2 warnings and 1 error generated.

```
</details>

Reply:
The output appears incorrect? Would you know why?

<details>
  <summary>Original</summary>

```console

---------------------------------------------------------------------------------------------------------
/* Matmul((4×4, u32), (4×4, u32), (4×4, u32), serial) */
alloc aa: (4×4, u32, L1) <- #1
  /* Matmul((4×4, u32), (4×4, u32, L1), (4×4, u32), serial) */
  tile (ab: (2×4, u32) <-[0, 2]- #0, ac: (2×4, u32) <-[0, 1]- #2)
    /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32), serial) */
    alloc ad: (2×4, u32, L1) <- ac
      /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32, L1), serial) */
      alloc ae: (2×4, u32, L1) <- ab
        /* Matmul((2×4, u32, L1), (4×4, u32, L1), (2×4, u32, L1), serial) */
        alloc af: (2×4, u32, RF)
          /* Move((2×4, u32, L1), (2×4, u32, RF), serial) */
          tile (ag: (1×1, u32, L1, ua) <-[0, 1]- ae, ah: (1×1, u32, RF, ua) <-[0, 1]- af)
            /* Move((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial) */
            ValueAssign(ag, ah)
          /* Matmul((2×4, u32, RF), (4×4, u32, L1), (2×4, u32, L1), serial) */
          tile (ai: (1×4, u32, RF, ua) <-[0, 2]- af, aj: (1×4, u32, L1, ua) <-[0, 1]- ad)
            /* Matmul((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, L1, ua), serial) */
            alloc ak: (1×4, u32, VRF, 4)
                /* Zero((1×4, u32, VRF, 4), serial) */
                VectorZero(ak)
                /* MatmulAccum((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, VRF, 4), serial) */
                tile (al: (1×1, u32, RF, ua) <-[0, 1]- ai, am: (1×4, u32, L1, ua) <-[1, 2]- aa)
                  /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, L1, ua), (1×4, u32, VRF, 4), serial) */
                  alloc an: (1×4, u32, VRF, 4)
                    /* Move((1×4, u32, L1, ua), (1×4, u32, VRF, ua, 4), serial) */
                    VectorAssign(am, an)
                    /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, VRF, 4), (1×4, u32, VRF, 4), serial) */
                    BroadcastVecMult(al, an, ak)
              /* Move((1×4, u32, VRF, 4), (1×4, u32, L1), serial) */
              VectorAssign(ak, aj)

#include <inttypes.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <arm_neon.h>
typedef uint32_t vui4 __attribute__ ((vector_size (4 * sizeof(uint32_t))));

__attribute__((noinline))
void kernel(
  uint32_t *restrict aa, 
  uint32_t *restrict ab, 
  uint32_t *restrict ac
) {
  for (int ad = 0; ad < 2; ad++) {
    uint32_t ae[8] __attribute__((aligned (128)));
    for (int af = 0; af < 2; af++) {
    for (int ag = 0; ag < 4; ag++) {
      ae[4 * af + ag] = aa[8 * ad + 4 * af + ag];
    }
    }
    for (int ah = 0; ah < 2; ah++) {
      vui4 aj;
      aj *= 0;  /* VectorZero */
      for (int ak = 0; ak < 4; ak++) {
        vui4 am;
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
        *(uint32x4_t *)(&aj) += ae[4 * ah + ak] * (*(uint32x4_t *)(&am)); /* BroadcastVecMult */
      }
      *(uint32x4_t *)(ac + 8 * ad + 4 * ah) = (*(uint32x4_t *)(&aj));  /* VectorAssign */
    }
  }
}

int main() {
  uint32_t buf0[16] __attribute__((aligned (128)));
  uint32_t buf1[16] __attribute__((aligned (128)));
  uint32_t buf2[16] __attribute__((aligned (128)));

  // Initialize the input tensors.
  for (int i = 0; i < 16; i++) {
    buf0[i] = i;
  }
  for (int i = 0; i < 16; i++) {
    buf1[i] = i;
  }

  kernel(&buf0[0], &buf1[0], &buf2[0]);

  // Print the output.
  for (int i = 0; i < 16; i++) {
    printf("%d ", buf2[i]);
  }

  return 0;
}

/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp3DiVJe/main.c:28:50: warning: incompatible pointer types assigning to 'const unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:8756:47: note: expanded from macro 'vld1q_u32'
  __ret = (uint32x4_t) __builtin_neon_vld1q_v(__p0, 50); \
                                              ^~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26817:21: note: expanded from macro 'vst1q_u32'
  uint32x4_t __s1 = __p1; \
                    ^~~~
/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp3DiVJe/main.c:28:19: warning: incompatible pointer types assigning to 'unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                  ^~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26818:26: note: expanded from macro 'vst1q_u32'
  __builtin_neon_vst1q_v(__p0, (int8x16_t)__s1, 50); \
                         ^~~~
2 warnings generated.

Output: 56 62 68 74 152 174 196 218 248 286 324 362 344 398 452 506 

```
</details>

<details>
  <summary>Proposed Code</summary>

```console

---------------------------------------------------------------------------------------------------------
/* Matmul((4×4, u32), (4×4, u32), (4×4, u32), serial) */
alloc aa: (4×4, u32, L1) <- #1
  /* Matmul((4×4, u32), (4×4, u32, L1), (4×4, u32), serial) */
  tile (ab: (2×4, u32) <-[0, 2]- #0, ac: (2×4, u32) <-[0, 1]- #2)
    /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32), serial) */
    alloc ad: (2×4, u32, L1) <- ac
      /* Matmul((2×4, u32), (4×4, u32, L1), (2×4, u32, L1), serial) */
      alloc ae: (2×4, u32, L1) <- ab
        /* Matmul((2×4, u32, L1), (4×4, u32, L1), (2×4, u32, L1), serial) */
        alloc af: (2×4, u32, RF)
          /* Move((2×4, u32, L1), (2×4, u32, RF), serial) */
          tile (ag: (1×1, u32, L1, ua) <-[0, 1]- ae, ah: (1×1, u32, RF, ua) <-[0, 1]- af)
            /* Move((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial) */
            ValueAssign(ag, ah)
          /* Matmul((2×4, u32, RF), (4×4, u32, L1), (2×4, u32, L1), serial) */
          tile (ai: (1×4, u32, RF, ua) <-[0, 2]- af, aj: (1×4, u32, L1, ua) <-[0, 1]- ad)
            /* Matmul((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, L1, ua), serial) */
            alloc ak: (1×4, u32, VRF, 4)
                /* Zero((1×4, u32, VRF, 4), serial) */
                VectorZero(ak)
                /* MatmulAccum((1×4, u32, RF, ua), (4×4, u32, L1), (1×4, u32, VRF, 4), serial) */
                tile (al: (1×1, u32, RF, ua) <-[0, 1]- ai, am: (1×4, u32, L1, ua) <-[1, 2]- aa)
                  /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, L1, ua), (1×4, u32, VRF, 4), serial) */
                  alloc an: (1×4, u32, VRF, 4)
                    /* Move((1×4, u32, L1, ua), (1×4, u32, VRF, ua, 4), serial) */
                    VectorAssign(am, an)
                    /* MatmulAccum((1×1, u32, RF, ua), (1×4, u32, VRF, 4), (1×4, u32, VRF, 4), serial) */
                    BroadcastVecMult(al, an, ak)
              /* Move((1×4, u32, VRF, 4), (1×4, u32, L1), serial) */
              VectorAssign(ak, aj)

#include <inttypes.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <arm_neon.h>
typedef uint32_t vui4 __attribute__ ((vector_size (4 * sizeof(uint32_t))));

__attribute__((noinline))
void kernel(
  uint32_t *restrict aa, 
  uint32_t *restrict ab, 
  uint32_t *restrict ac
) {
  for (int ad = 0; ad < 2; ad++) {
    uint32_t ae[8] __attribute__((aligned (128)));
    for (int af = 0; af < 2; af++) {
    for (int ag = 0; ag < 4; ag++) {
      ae[4 * af + ag] = aa[8 * ad + 4 * af + ag];
    }
    }
    for (int ah = 0; ah < 2; ah++) {
      vui4 aj;
      aj *= 0;  /* VectorZero */
      for (int ak = 0; ak < 4; ak++) {
        vui4 am;
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
        *(uint32x4_t *)(&aj) += ae[4 * ah + ak] * (*(uint32x4_t *)(&am)); /* BroadcastVecMult */
      }
      *(uint32x4_t *)(ac + 8 * ad + 4 * ah) = (*(uint32x4_t *)(&aj));  /* VectorAssign */
    }
  }
}

int main() {
  uint32_t *restrict an;
  posix_memalign((void **)&an, 128, 16*sizeof(uint32_t));
  memset(an, 0, 16*sizeof(uint32_t));
  uint32_t *restrict ao;
  posix_memalign((void **)&ao, 128, 16*sizeof(uint32_t));
  memset(ao, 0, 16*sizeof(uint32_t));
  uint32_t *restrict ap;
  posix_memalign((void **)&ap, 128, 16*sizeof(uint32_t));
  memset(ap, 0, 16*sizeof(uint32_t));

  kernel(an + 0, ao + 0, ap + 0);

  // Print the output.
  for (int i = 0; i < 16; i++) {
    printf("%d ", ap[i]);
  }

  // Free the buffers.
  free(an);
  free(ao);
  free(ap);

  return 0;
}

/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp93fumG/main.c:28:50: warning: incompatible pointer types assigning to 'const unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:8756:47: note: expanded from macro 'vld1q_u32'
  __ret = (uint32x4_t) __builtin_neon_vld1q_v(__p0, 50); \
                                              ^~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26817:21: note: expanded from macro 'vst1q_u32'
  uint32x4_t __s1 = __p1; \
                    ^~~~
/var/folders/l3/vsszss2s3p98jhhvrn54kwhm0000gn/T/.tmp93fumG/main.c:28:19: warning: incompatible pointer types assigning to 'unsigned int *' from 'uint32x4_t *' [-Wincompatible-pointer-types]
        vst1q_u32((uint32x4_t *)(&am), vld1q_u32((uint32x4_t *)(ab + 4 * ak)));  /* VectorAssign */
                  ^~~~~~~~~~~~~~~~~~~
/opt/homebrew/Cellar/llvm/16.0.6/lib/clang/16/include/arm_neon.h:26818:26: note: expanded from macro 'vst1q_u32'
  __builtin_neon_vst1q_v(__p0, (int8x16_t)__s1, 50); \
                         ^~~~
2 warnings generated.

Output: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 

```
</details>

Reply:
The generated code is no longer filling in the buffers with sequential integers, so I think the output is correct.

For context, your implementation differs from the Python implementation, which does one of three things, depending on parameters given by the caller: leaves the buffers uninitialized, fills them with `rand()`, or [fills them with data given by the caller](https://github.com/samkaufman/morello/blob/4aa8d35c945a6133d1663960e7f26c29487c2639/python/morello/codegen/gen.py#L650C13-L658).

I'm not sure what's up with the ARM-specific compiler warnings.

Reply:
I see. Thank you.

> I'm not sure what's up with the ARM-specific compiler warnings.

This can be ignored for now since this also happened in the Python implementation IIRC.